### DISCUSSION TITLE
GitLab CI, PCH, clang, changing build dir, absolute paths in cmake_pch.hxx.pch

### DISCUSSION THREAD

**Discussion by samopolacek** [ccache/ccache#1368] [Support]
I am trying to make ccache work in our GitLab pipelines. 

I was able to make ccache work in almost the whole matrix I am trying it on (Clang, GCC, debug, release, PCH, no PCH).

The only issue is combination of precompiled header, Clang,  and changing build directory.

Our build directory is unique for each commit (`GIT_CLONE_PATH: ${CI_BUILDS_DIR}/${CI_PROJECT_ID}_${CI_COMMIT_SHA}/`).
I am using `CCACHE_SLOPPINESS=time_macros,pch_defines,include_file_mtime,include_file_ctime` and `-Xclang -fno-pch-timestamp` (for Clang only),  and `CCACHE_BASEDIR: "${CI_PROJECT_DIR}"`, `CCACHE_NOHASHDIR: "true"`.
However, ccache still failed to hit for precompiled headers. What I found out is that all of the produced `build/**/CMakeFiles/*.dir/cmake_pch.hxx.pch` files contain many absolute paths to the build directory (checked for Clang).

Example of `strings cmake_pch.hxx.pch  | grep /build | head -n 10`

```
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/src/SomeLib/CMakeFiles/SomeLibSomeLib.dir/cmake_pch.hxx.cxx
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/src/SomeLib/CMakeFiles/SomeLib.dir
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/src/SomeLib/CMakeFiles/SomeLib.dir/cmake_pch.hxx
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/src/SomeLib/CMakeFiles/SomeLib.dir/cmake_pch.hxx.cxx
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/src/SomeLib/precompiled.h
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/../.conan_cache/grpc/1.50.2/sky/stable/package/0efde2e33e6c4dc257142ac4c61c4e5ac4c09546/include/grpcpp/grpcpp.h
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/../.conan_cache/grpc/1.50.2/sky/stable/package/0efde2e33e6c4dc257142ac4c61c4e5ac4c09546/include/grpc/grpc.h
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/../.conan_cache/grpc/1.50.2/sky/stable/package/0efde2e33e6c4dc257142ac4c61c4e5ac4c09546/include/grpc/support/port_platform.h
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/../.conan_cache/grpc/1.50.2/sky/stable/package/0efde2e33e6c4dc257142ac4c61c4e5ac4c09546/include/grpc/impl/codegen/port_platform.h
/builds/24211_60d144f0933884836b7fb0061ddb53451824feeb/build/../.conan_cache/grpc/1.50.2/sky/stable/package/0efde2e33e6c4dc257142ac4c61c4e5ac4c09546/include/grpc/byte_buffer.h
```
When I look into the ccache log file, I see: "No <hash> in local storage", "Not considering cached precompiled header in preprocessor mode", "Running real compiler".

Comparison of GCC and Clang `ccache -v --show-stats`

```
# Clang
$ ccache -v --show-stats
Cache directory:    /builds/24211_fca966f825df183719066aa5ce67faa10b959b40/.ccache
Config file:        /builds/24211_fca966f825df183719066aa5ce67faa10b959b40/.ccache/ccache.conf
System config file: /etc/ccache.conf
Stats updated:      Thu Dec 14 12:15:12 2023
Cacheable calls:     282 /  282 (100.0%)
  Hits:                0 /  282 ( 0.00%)
    Direct:            0
    Preprocessed:      0
  Misses:            282 /  282 (100.0%)
Successful lookups:
  Direct:              0 /  282 ( 0.00%)
  Preprocessed:        0 /  282 ( 0.00%)
Local storage:
  Cache size (GB):  0.46 / 5.00 ( 9.13%)
  Files:            2876
  Hits:                0 /  278 ( 0.00%)
  Misses:            278 /  278 (100.0%)
  Reads:             560
  Writes:            564
```

```
# GCC
$ ccache -v --show-stats
Cache directory:    /builds/24211_9291a2e24e73e1d47f2b5cc5e83b36b93ae14b10/.ccache
Config file:        /builds/24211_9291a2e24e73e1d47f2b5cc5e83b36b93ae14b10/.ccache/ccache.conf
System config file: /etc/ccache.conf
Stats updated:      Thu Dec 14 11:08:11 2023
Cacheable calls:     282 /  282 (100.0%)
  Hits:              282 /  282 (100.0%)
    Direct:          278 /  282 (98.58%)
    Preprocessed:      4 /  282 ( 1.42%)
  Misses:              0 /  282 ( 0.00%)
Successful lookups:
  Direct:            278 /  282 (98.58%)
  Preprocessed:        4 /    4 (100.0%)
Local storage:
  Cache size (GB):  0.14 / 5.00 ( 2.77%)
  Files:             581
  Hits:              282 /  282 (100.0%)
  Misses:              0 /  282 ( 0.00%)
  Reads:             564
  Writes:              4

# 4 PCH for each build directory change, i.e. when run with a different commit
```

Has anybody encountered the same issue?

----
Edit: Add info about CCACHE_BASEDIR, CCACHE_NOHASHDIR
Edit: Add sample of `strings cmake_pch.hxx.pch`
Edit: Add info from ccahce log file
Edit: Add gcc vs. clang ccache stats

**Comment by KilianBl** [comment]
Same problem here. Is there a solution? @samopolacek @jrosdahl 

Thanks in advance

**Reply by samopolacek** [reply]
@KilianBl disable precompiled headers for configuations that keep missing.

The "drawback" is that you have to fix missing includes, which is a good thing, but requires the work to be done.

I did that by conditionally (if running in a CI  pipeline and is Clang) overriding the target_precompile_headers CMake function and made it do nothing.

**Reply by KilianBl** [reply]
Thanks for getting back to me, @samopolacek. We were hoping to avoid stopping using PCH with Clang and compiling in different directories, but we know it's an option.

Maybe we should open an issue to get some feedback on this @jrosdahl? I couldn't find anything in the docs saying it's not supported.

**Reply by jrosdahl** [reply]
If you have a simple example that demonstrates a problem, you can open an issue if you like. That being said, I don't use precompiled headers myself so I likely won't have much feedback to give. It's possible that it used to work better with older Clang versions, I don't know.

**Reply by KilianBl** [reply]
Issue created https://github.com/ccache/ccache/issues/1563

Thanks
